rest-api/main.go

```diff
@@ -0,0 +1,195 @@
+	Designation string `json:"designation"`
+}
+
+var employees = []Employee{
+	{
+		ID:          "464eee0b-802f-4193-8405-c5f39c1ecd72",
+		Name:        "John Doe",
+		Email:       "john.doe@dummy.com",
+		Active:      true,
+		CreatedAt:   1720430892,
+		UpdatedAt:   1720430892,
+		Designation: "Software developer",
+	},
+	{
+		ID:          "c5f12d01-8113-4275-9a2a-77a1c40690f4",
+		Name:        "James Bond",
+		Email:       "bond007@dummy.com",
+		Active:      true,
+		CreatedAt:   1720430892,
+		UpdatedAt:   1720430892,
+		Designation: "Software developer",
+	},
+}
+
+// GET all employee function
+func getAllEmployees(context *gin.Context) {
+	context.JSON(http.StatusOK, employees)
+}
+
+// GET employee by id function
+func getEmployeeByID(context *gin.Context) {
+	id := context.Param("id")
+	for _, emp := range employees {
+		if emp.ID == id {
+			context.JSON(http.StatusOK, emp)
+			return
+		}
+	}
+	context.JSON(http.StatusNotFound, gin.H{"error": "Employee not found"})
+}
+
+// POST employee function
+func postEmployees(context *gin.Context) {
+	var newEmployee Employee
+
+	if err := context.BindJSON(&newEmployee); err != nil {
+		context.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	newEmployee.ID = uuid.New().String()
+
+	for _, emp := range employees {
+		if emp.ID == newEmployee.ID {
+			context.JSON(http.StatusConflict, gin.H{"error": "Employee with same ID already exists"})
+			return
+		}
+	}
+
+	currentTime := getCurrentTimestamp()
+	newEmployee.CreatedAt = currentTime
+	newEmployee.UpdatedAt = currentTime
+
+	employees = append(employees, newEmployee)
+	context.JSON(http.StatusCreated, employees)
+}
+
+// PATCH employee data function
+func patchEmployee(context *gin.Context) {
+	id := context.Param("id")
+	var patchedEmployee Employee
+
+	if err := context.BindJSON(&patchedEmployee); err != nil {
+		context.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	for i, emp := range employees {
+		if emp.ID == id {
+			if emp.Active == false {
+				context.JSON(http.StatusBadRequest, gin.H{"error": "Employee is not active and cannot be updated"})
+				return
+			}
+
+			if patchedEmployee.Name != "" {
+				employees[i].Name = patchedEmployee.Name
+			}
+			if patchedEmployee.Email != "" {
+				employees[i].Email = patchedEmployee.Email
+			}
+			if patchedEmployee.Designation != "" {
+				employees[i].Designation = patchedEmployee.Designation
+			}
+
+			employees[i].UpdatedAt = getCurrentTimestamp()
+
+			context.JSON(http.StatusOK, employees[i])
+			return
+		}
+	}
+
+	context.JSON(http.StatusNotFound, gin.H{"error": "Employee not found"})
+}
+
+// PUT employee data function
+func putEmployee(context *gin.Context) {
+	id := context.Param("id")
+	var updatedEmployee Employee
+
+	if err := context.BindJSON(&updatedEmployee); err != nil {
+		context.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
+		return
+	}
+
+	for i, emp := range employees {
```

Comment:
Business logic shouldn't be part of controller. move such logic to service layer by introducing another struct `EmployeeService`.

Reply:
A controller should only be responsible to route request/response to/from service layer.